src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.SymbolToProjectId.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// Contains caches to help us quickly and accurately map from <see cref="ISymbol"/>s to <see
+        /// cref="ProjectId"/>.
+        /// </summary>
+        /// <remarks>
+        /// Note: this is not <see langword="readonly"/> as this is a struct and will mutate as operations are called on
+        /// it.  However, in the common case where this data is not needed, it will have not have any impact.
+        /// </remarks>
+        private SymbolToProjectId _symbolToProjectId;
+
+        /// <inheritdoc cref="Solution.GetExactProjectId"/>
+        public ProjectId? GetExactProjectId(ISymbol symbol)
+            => _symbolToProjectId.GetExactProjectId(symbol);
+
+        /// <summary>
+        /// A helper type that can be used to map from symbols to a project they could have come from.
+        /// </summary>
+        private struct SymbolToProjectId
+        {
+            /// <summary>
+            /// Weak cache from a <see cref="Compilation"/> to the ID of the <see cref="Project"/> it was created from.
+            /// Only used for mapping a <see cref="Project"/>'s <see cref="Compilation.GlobalNamespace"/> back.  These
+            /// <see cref="INamespaceSymbol"/>'s do not belong to a <see cref="IAssemblySymbol"/> or <see
+            /// cref="IModuleSymbol"/> and thus cannot be tracked with <see
+            /// cref="CompilationTracker.State.CompilationAssembliesAndModules"/>.
+            /// </summary>
+            private static readonly ConditionalWeakTable<Compilation, ProjectId> s_compilationToProjectId
+                = new ConditionalWeakTable<Compilation, ProjectId>();
+
+            private readonly SolutionState _solutionState;
+
+            /// <summary>
+            /// Cache we use to map between assembly and module symbols and the project they came from.  That way if we
+            /// are asked about many symbols from the same assembly/module we can answer the question quickly after
+            /// computing for the first one.
+            /// </summary>
+            private ConditionalWeakTable<ISymbol, ProjectId?>? _assemblyOrModuleSymbolToProjectId;
```

Comment:
We already have the existing s_assemblyOrModuleSymbolToProjectMap which was the problematic one for you. Can we just make that existing one hold onto a tuple of the two ProjectIds rather than having two ConditionalWeakTables entirely? That'll be cheaper than having two maps no matter what I'd think, and probably makes for less GC walking?

Reply:
i'm happy to make such a change  in tihs PR.  once we get it working :D